src/Form/ConstraintForm.php

```diff
@@ -0,0 +1,233 @@
+   * @param \Drupal\password_enhancements\PasswordConstraintPluginManager $password_constraint_plugin_manager
+   *   Password constraint plugin manager.
+   * @param \Drupal\Core\Messenger\MessengerInterface $messenger
+   *   Messenger.
+   */
+  public function __construct(ConfigEntityStorageInterface $constraint_config_entity_storage, PasswordConstraintPluginManager $password_constraint_plugin_manager, MessengerInterface $messenger) {
+    $this->constraintConfigEntityStorage = $constraint_config_entity_storage;
+    $this->passwordConstraintPluginManager = $password_constraint_plugin_manager;
+    $this->messenger = $messenger;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container): ConstraintForm {
+    return new static(
+      $container->get('entity_type.manager')->getStorage('password_enhancements_constraint'),
+      $container->get('plugin.manager.password_constraint'),
+      $container->get('messenger')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function form(array $form, FormStateInterface $form_state): array {
+    $form = parent::form($form, $form_state);
+
+    $form['#title'] = $this->t('Edit constraint');
+
+    $options = [];
+    foreach ($this->passwordConstraintPluginManager->getDefinitions() as $definition) {
+      $options[$definition['id']] = $definition['name'];
+    }
+
+    $form['policy'] = [
+      '#type' => 'hidden',
+      '#value' => $this->getRequest()->get('password_enhancements_policy')->id(),
+    ];
+
+    $settings_wrapper_id = 'ajax-wrapper';
+    $form['type'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Constraint'),
+      '#description' => $this->t('The type of the constraint.'),
+      '#options' => $options,
+      '#default_value' => $this->entity->getType(),
+      '#required' => TRUE,
+      '#ajax' => [
+        'callback' => '::updateForm',
+        'event' => 'change',
+        'wrapper' => $settings_wrapper_id,
+      ],
+      '#disabled' => !$this->entity->isNew(),
+    ];
+
+    $form['wrapper'] = [
+      '#type' => 'container',
+      '#prefix' => "<div id='{$settings_wrapper_id}'>",
+      '#suffix' => '</div>',
+      'descriptionSingular' => [
+        '#type' => 'textfield',
+        '#title' => $this->t('Description (singular)'),
+        '#description' => $this->t('The singular description that should be displayed for the password field.'),
+        '#required' => TRUE,
+      ],
+      'descriptionPlural' => [
+        '#type' => 'textfield',
+        '#title' => $this->t('Description (plural)'),
+        '#description' => $this->t('The plural description that should be displayed for the password field.'),
+        '#required' => TRUE,
+      ],
+      'settings' => [
+        '#type' => 'fieldset',
+        '#title' => $this->t('Constraint settings'),
+        '#tree' => TRUE,
+      ],
+    ];
+
+    try {
+      $form['wrapper'] = static::updateForm($form, $form_state);
```

Comment:
You are calling a non-static function as static.

Reply:
Weird, it is not marked as a warning/notice/error by php neither by phpstorm 😮